### ISSUE TITLE
`discover` using dist-git source should apply downstream patches before running tests

### ISSUE THREAD

**Issue by thozza** [teemtee/tmt#1482]
Using the following Plan:
```
summary: Run unit tests from source
discover:
    how: fmf
    dist-git-source: true
execute:
    how: tmt
```
results in TMT not taking downstream patches into consideration and not applying them at all on the extracted sources before executing tests. However downstream patches may be sometimes required for the unit tests to pass in the first place.

If tests are run on sources without applying downstream patches first, then as a result they are run on a version of sources which is not used anywhere (even not for building RPMs). TMT should really use the available RPM tooling and instead run something like `***pkg prep` and take the result (extracted archive + applied patches) and execute tests on such sources.

The relevant code is https://github.com/teemtee/tmt/blob/26fe39b82b94dea366073e934a6c1f77b797d7d9/tmt/steps/discover/__init__.py#L274-L305

**Comment by psss** [comment]
@lukaszachy, what were the concerns related to executing the `prep` step of the specfile? We would need to bring some extra dependencies (which should be relatively simple) but there were some more...

**Comment by lukaszachy** [comment]
a) We need to respect _target_ build macros (e.g. host is fedora but SUT will be rhel)
b) Shouldn't allow any malicious spec file change host machine (afaik in %prep one put do anything).

As a workaround one should be able to run rpmbuild in 'prepare' and let it replace current code. Problem is in push/pull might replace data (not sure) and it will still not detect _new_ tests.

**Comment by thozza** [comment]
Thanks for describing reasons for the current behavior.

> a) We need to respect _target_ build macros (e.g. host is fedora but SUT will be rhel) 

Not sure I understand this point and I'm not sure how valid it is. My expectation would be that the test infrastructure will use the same OS on the host as is the SUT. What would be the point from the PoV of the validity of results to run RHEL tests on Fedora?

> b) Shouldn't allow any malicious spec file change host machine (afaik in %prep one can put/do anything).

`%prep` does not need to be run as a privileged user. It can be run in a container, etc. The CI system can restrict if tests should be run automatically for "untrusted" contributors. This is commonly available on e.g. GitHub. Additionally, even the code that is executed as part of the test and comes from the upstream sources may be malicious.

> As a workaround one should be able to run rpmbuild in 'prepare' and let it replace current code. Problem is in push/pull might replace data (not sure) and it will still not detect _new_ tests (read: those added/deleted by patch).

I'm sure we can workaround this by executing whatever we need (even the `%prep`) in the test scrip and not use `dist-git-source: true`. However this makes the option literally useless in a general case, because one can not use it reliably if the SPEC file applies any patches. Applying patches is quite common for RHEL packages where changes are usually backported as patches.

The current behavior of the option makes it impossible to run e.g. unit tests on the same sources as those that are used for building the actual RPMs, but its name IMO creates a false impression that it is so.

**Comment by psss** [comment]
> Not sure I understand this point and I'm not sure how valid it is. My expectation would be that the test infrastructure will use the same OS on the host as is the SUT. What would be the point from the PoV of the validity of results to run RHEL tests on Fedora?

The `discover` step part (which takes care of sources extraction) is done on the test runner where `tmt` command is executed and that one does not have to be the target guest as you can use one of the `provision` methods to safely execute tests from your laptop in a `container` or a `virtual` machine. Also in the Testing Farm `provision --connect` is used to execute tests.

**Comment by psss** [comment]
> However this makes the option literally useless in a general case, because one can not use it reliably if the SPEC file applies any patches.

This applies only for patches which are modifying tests, so not completely useless. The application under test is, of course, built with all patches applied.

**Comment by thozza** [comment]
> > However this makes the option literally useless in a general case, because one can not use it reliably if the SPEC file applies any patches.
> 
> This applies only for patches which are modifying tests, so not completely useless. The application under test is, of course, built with all patches applied.

In our scenario, an external dependency caused a unit test to fail and a patch was fixing an existing unit test. In our case (osbuild-composer), we are the upstream and we were able to quickly make a new release, however for any external project, this would not be possible. AFAICT, there would be no way to workaround this limitation without a new release, because the sources are fetched from the URL extracted from the SPEC file (at least from the brief look I had at the code).

Personally, I think that the situation is not really ideal even if the end result was that new tests added by a patch are not executed. The use of such test executor looses its value as a result of this behavior, especially if used to gate changes landing in a distribution, because new backported functionality is not tested.

**Comment by lukaszachy** [comment]
I completely agree that tmt should apply patches (however safely and respecting SUT which would be "fun" to discover but dist-git branch should provide initial idea). Current behaviour was the first step to have something right now with lets make it better in the future.

**Comment by thozza** [comment]
So what is the plan here? Is there a plan / desire to support applying downstream patches when using dist-git source? Or should we stop using it at all?

I still think that the current approach is unfortunate, since it is potentially testing different sources than what is really being built from dist-git. Also the implementation limitation is making it impossible to fix CI / Gating failures in unit tests by backporting the fix as a downstream patch.

**Comment by lukaszachy** [comment]
Unfortunately priorities were elsewhere during the past quarters. Thank you for raising the questions and thus the priority to solve this.
Currently the best approach seems to be to use mock to prepare the interim environment (mainly to have correct rpm macros installed) and run '%prep' phase. 

More precise plan should be ready by end of this or early next week.

**Comment by lukaszachy** [comment]
Sorry for late response. We had a chat about possibilities and way to go seems to create new step which runs before the discover. This `begin` step (no idea about the final name yet, pre-discover will be 3rd step with 'pr' prefix) will reuse configuration of `provision` step to get the same SUT and will run `rpmbuild -bp` on such machine and later pull patched BUILD directory which will be used for discovery.  

Another considered approach was to use 'mock' but IMO using it would be more difficult to map to specific compose which might be necessary for some cases (and user / CI is expected to provide correct provision configuration). Otherwise one would need to either guess the mock config from provision config (which will not work for 'connect') or make it explicit for the discover step. Another mock drawback is the need for user to be in 'mock' group.

So the plan for now is:
- Create new step (for now called 'begin', better name really appreciated)
- Step will peek into discover config and run only if `dist-git-source` is enabled (for now, possible ideas to utilize this step in the future)
- Step will use provision configuration (opt-in option to reuse provisioned machine, for now there won't be way to provide different 'provision -h connect' config for 'begin' step)
- Initial version will not run 'prepare' (but might in the future as opt-in, reason: enable crb, install builds, ...)
- Once 'rpmbuild -bp' finishes successfully, BUILD directory will be pulled from the guest and 'discover' step will continue from obtained data. Machine will be discarded (in the future it might be requested to be reused).

**Comment by psss** [comment]
One quick brainstorm from a devconf chat: What about just doing the `rpmbuild -bp` part in a container? In order to make it working we could detect the target distro from the `context`. Sounds probably a bit like a hack, but on the other hand, the implementation would not be that complex. If `distro` is not provided, we could fall back to the old behaviour. Thoughts?

**Comment by lukaszachy** [comment]
> One quick brainstorm from a devconf chat: What about just doing the `rpmbuild -bp` part in a container? In order to make it working we could detect the target distro from the `context`. Sounds probably a bit like a hack, but on the other hand, the implementation would not be that complex. If `distro` is not provided, we could fall back to the old behaviour. Thoughts?

Had similar idea but I ran into two problems:
1. podman (or docker or..) dependency will become mandatory
2. some container images don't exist or are super difficult to find.  E.g. `tmt -c distro=rhel-9.4 run -a provision -h minute -i rhel-9.4` - under both 'rhel-9.4' is hidden some specific compose and I'm not aware we have usable rhel containers (even internally one has to inject private dnf repositories as the expectation is to use subscription-manager in the image)

Unfortunately reusing current provision step configuration ( single machine) and letting user to modify that if necessary seems to be the correct way. However too difficult.

I'm starting to think that allowing tmt to rediscover tests after prepare step might be better approach, because one might need to run prepare step also to be able to run rpmbuild (at least somehow  enable codeready repo)

**Comment by psss** [comment]
Hm... so what about inserting a little `rediscover` step after `prepare` if `dist-git-source` detected? Would it be too bad to do the source preparation on the guest itself? Without any cleanup?

**Comment by thozza** [comment]
Note that I realized another issue with the approach used by TMT, which is related to this one. TMT downloads the upstream sources from the URL in the SPEC file. However, to be on the safe side, it should really download sources from the dist-git lookaside cache. While it is not usual, it happened that some upstreams retrospectively changed release tarballs. The only 100% correct approach is to download sources from the respective lookaside cache and then run the whole `%prep` phase from SPEC. Only this guarantees that the sources are really the ones that will be (or were) used to build RPMs.

On a different note, I managed to implement a solution of my own to prep sources "the right way", which works both in CentOS Stream CI as well as in RHEL CI (and should also in Feodra, but I didn't test it there). For future reference, this is the relevant part:
https://gitlab.com/redhat/centos-stream/rpms/osbuild/-/blob/90ab1b7b0ca2f7449fa0203ee260a085f34bfcfe/tests/unit-tests/run-unit-tests.sh#L6-63

**Comment by happz** [comment]
As discussed, @lukaszachy will write a bit more about how this would be implemented and how big are the worms in this can.

**Comment by lukaszachy** [comment]
The promised explanation why this takes ages:

For every provision method except 'how: local' the tmt's attempts to be safe to execute any plan without the risk compromising the host system. No packages are installed on the host, no scripts are run,  'git clone' during the discover and 'rsync' workdir back to the host are the only things which happen. I'd say we should keep this promise valid for dist-git as well. The 'rpm-build -bp' can do a lot of things and might require several other dependencies tmt could compromise the system if it was allowed to run. The other feature which would not work is the 'cross-distro' - the host (where 'tmt run' is exexuted) and guest (where tests run) can be completely different distributions (e.g. using fedora host to run tests against ancient rhel-6 guest).

Thus the idea was to use some guest to be able to run 'rpmbuild -bp' on it. But how to get that guest? In the most common TF usage guests are 'how: connect' so tmt has no way to spawn another instance on demand. Re-using (selected) guest could be an option but the flow would change significantly as 'prepare' might need to run as well (enable necessary repositories and/or install build dependencies as rpmbuild and rpm macros). Thus the flow would be: during discover's dist-git:true run selected provision and prepare steps, then do the 'rpmbuild -bp' and pull sources back to host. 
The other motivation to reuse same guest configuration is to have same versions of macros available (to be sure that spec file is processed correctly). This means that using 'mock' or 'container' are out of the options as such their configuration/images might not be available.

There are also two discover plugins. The 'how: shell' is easier - the list of tests can be defined only in the plan so all what 'dist-git: true' can do is prepare sources. But no new tests (from the tmt point of view) can appear. No re-discover for tests is necessary, only to have up-to-date worktree available

But 'how: fmf' needs to do 'tmt tests ls' on extracted sources to fetch the list of test metadata as well as their sources.
Meaning that once tmt applies patches, it should pull the extracted files back to the host (because tmt doesn't need to be present on guest) and finally discover/filter tests and then install require/recommend again (new dependencies might appear)

However this 'postponed' discover will bring troubles to '--dry' and running just selected steps'. E.g. `tmt run discover` cannot be completed without 'provision and 'prepare' running.

What do you think about 'implicit' prepare phase for 'rpmbuild -bp && rsync && discover' running just after 'recommend' installation? In the case of 'how: fmf' and when new require/recommend appear it will also do the require/recommend install again. 
In the case of multi-host (plan with multiple provision phases) - How tmt should pick machine where rpm-build will run? Or should it run everywhere? Should it be (optional) flag? We could use 'where' to allow user to mark such machine (without this value it will run everywhere)

**Comment by psss** [comment]
> What do you think about 'implicit' prepare phase for 'rpmbuild -bp && rsync && discover' running just after 'recommend' installation? In the case of 'how: fmf' and when new require/recommend appear it will also do the require/recommend install again.

Adding an implicit `prepare` step phase sounds good to me. The `provision` step is already complete which means the guest could be used for the `rpmbuild` stuff. Why do you propose to go only after installation of recommended packages? Couldn't we install just build requires based on the spec (which should be correctly set), do the `rpmbuild -bp & rsync & rediscover` and only then continue with the installation of required and recommended packages?

**Comment by lukaszachy** [comment]
> Why do you propose to go only after installation of recommended packages? Couldn't we install just build requires based on the spec (which should be correctly set)

Just for 'implicit requires'. IIRC the 'gcc' and friends are supposed to be already present on the build systems (but languages usually do not need them).  I'm not fan of having them installed on every system (eg. most of python packages doesn't need those).

**Comment by lukaszachy** [comment]
Got answer from more seasoned packager about 'packages assumed to be present on build system':
On Fedora: `dnf install @buildsys-build` 
On RHEL it's enumerated in `/etc/mock/templates/rhel-8.tpl` as `config_opts['chroot_setup_cmd'] = 'install tar gcc-c++ redhat-rpm-config redhat-release which xz sed make bzip2 gzip gcc coreutils unzip diffutils cpio bash gawk rpm-build info patch util-linux findutils grep'`

**Comment by thrix** [comment]
From discussion:
* for multihost just pick a guest, do not overcomplicate

**Comment by lukaszachy** [comment]
It took ages, but if you are still interested in this feature on tmt side, @thozza your feedback for #2641 will be very appreciated.

**Comment by thozza** [comment]
Unfortunately, I'm no longer interested.

I workarounded the problem in our tests and have no plan to touch it until it keeps working. FWIW, I had a brief look at the PR and the change looks like it should cover our initial request.
